### ISSUE TITLE
design keyboard help dialog

### ISSUE THREAD

**Issue by emily-phet** [phetsims/ratio-and-proportion#82]
this sim has many different UI features:
- sliders
- custom two-hand interaction
- custom jump keys (numbers and letter 'j')
- radio button group
- accordion box
- combobox
- number spinner

Once we complete at least one reasonably successful round of interviews with screen reader users, we need to design the keyboard help dialog.

**Comment by emily-phet** [comment]
@Matthew-Moore240 - can you please draft text for the keyboard help dialog text? Let me know if you'd like to discuss first.

**Comment by Matthew-Moore240** [comment]
Will do!

**Comment by emily-phet** [comment]
@Matthew-Moore240 Let me know if you have any questions on this. We'll need this complete (drafted, iterated on, finalized) by Sept. 15th.

**Comment by terracoda** [comment]
Here are examples from sims with similar interactions:

Sliders & Combobox - Molarity
![molarity-kbs](https://user-images.githubusercontent.com/1902116/92239881-477b7600-ee96-11ea-811c-ce4820e6eae9.png)

Sliders, Number Spinners, Custom Grab & Move Ruler & Custom Jump keys - GFL
![gfl-kbs](https://user-images.githubusercontent.com/1902116/92240059-93c6b600-ee96-11ea-8371-8c911cfbeb16.png)

Custom Grab & Move & Jump Keys - BASE & Friction
![base-kbs](https://user-images.githubusercontent.com/1902116/92240385-1d768380-ee97-11ea-9c89-dc2d708613d0.png)
![Screen Shot 2020-09-04 at 10 12 44 AM](https://user-images.githubusercontent.com/1902116/92240426-32531700-ee97-11ea-8785-7fcc18c422ec.png)

Accordion Box - No examples yet

Basic Actions section covers:
- Tabbing, Group Navigation, Buttons, Checkboxes, and Radio Buttons

**Comment by terracoda** [comment]
@Matthew-Moore240, you'll need to also look at the A11y views as the wording is slightly different when the instructions are spelled out.

**Comment by terracoda** [comment]
@Matthew-Moore240, I polished up the draft content for the Keyboard Shortcuts dialog. We can discuss at Tuesday's meeting or if you like it, just assign it to @zepumph.

This sim has different actions on different screens, so 2 different dialogs will be needed.

Link to [Keyboard Shortcuts Dialog](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit#heading=h.sd1v4vya1mk3) in Design Doc

**Comment by zepumph** [comment]
After discussing with @jessegreenberg, we are going to add support for screen-specific help content over in https://github.com/phetsims/joist/issues/668. I'll do that before this issue.

**Comment by zepumph** [comment]
@terracoda , here are some notes to review:
* I used the built in slider keyboard help component to get to where we are for the 'Move Left or Right Hand' section. If we need to change individual rows to be more hand-specific, we can, but that will be a change to the common code component, so I thought I would check. Perhaps we can solve this by changing the heading to 'Move Left or Right Hand Individually'
* Should there be periods at the end of description sentences? I see both w/o often.
*For "Steps to Change a Challenge" section: I created a new general section for explaining how to use a combo box. In it things are a bit easier if we can have just one word used consistently, instead of switching between "ratio challenge" and "challenge", can we just pick one to use for all rows? The heading will still have a different entry for "Challenge"; this is just for the rows of the section.
* I didn't understand "set" here, "Move through set challenges." What about instead "Move through possible ratio challenges?"
* For "Steps to Change a Challenge" (really for combo box instructions in general), I made the spacing a bit less. I liked this because it looked like one set of instructions. Let me know if you don't like that though. 


Here are the two dialogs:
Discover Screen:
![image](https://user-images.githubusercontent.com/6856943/95149317-13fc6780-0732-11eb-95f0-6d6d0a9e33c4.png)

Create Screen: 
![image](https://user-images.githubusercontent.com/6856943/95149346-21195680-0732-11eb-81e3-ce38ad4e7b20.png)

**Comment by terracoda** [comment]
@zepumph, here's a better example from GFLB where we have left/right sliders and up/down number pickers, and where the visual and physical interaction of moving left/right or up/down is kind of strong. Of course, with sliders, all the arrow keys work. That said, I think we have the choice to pick a direction that is relevant to the sim. I think this is what we did for GFLB.

![Screen Shot 2020-10-06 at 1 54 10 PM](https://user-images.githubusercontent.com/1902116/95230100-fa175f00-07db-11eb-9a89-56ba72eaaeef.png)

**Comment by terracoda** [comment]
@zepumph, I have made some notes in the design doc that we can discuss at today's meeting.

**Comment by terracoda** [comment]
@zepumph, and I should say, the samples are a great start. It's great to see they content laid out.

**Comment by zepumph** [comment]
Today during design meeting, @terracoda @Matthew-Moore240 and I went through the keyboard help content and made many changes to the design doc. @terracoda will clean it up and assign me when it is ready for a direct copy.

Part of this will be adding options to the slider help text to improve its verbage.

**Comment by terracoda** [comment]
@zepumph, the content for the Keyboard Shortcuts dialog is updated and ready for implementation.
Use this link to jump to the sections with the [content for both dialogs](https://docs.google.com/document/d/1HJvvGWYaUv9tlat9S7omQQVng8fGB4FuUhTEM1qx1z8/edit#heading=h.6y493yuizr6z).

**Comment by zepumph** [comment]
Everything should either be implemented, or have a comment in the design doc about implementation. I believe the only main thing left is to determine how we want to set up the common code combo box help section header.

Desired text for RaP:
"Choose a Challenge Ratio"

Current text for molarity:
"Steps to Change Solute"

If there was a single sentence that could easily cover all cases, like "Choose a {{filledInThing}}", then maybe that would be preferred. Alternatively, we could have each usage of the combo box help section provide a string of the whole header (with a default for the option). This second idea is what we do for slider. Actually, now that I hear myself typing it out. I think that is the best way to implement the code, and it is the most flexible. I'll make the change now.

**Comment by zepumph** [comment]
Ok. Now I like the way that the common code combo box help section heading option is set up. I think that RaP is totally done, save a couple things in design doc comments. @terracoda please review by looking first at these comments, and then at the design doc comments (I think that will make the most sense).

I will make a note over in https://github.com/phetsims/molarity/issues/234 to discuss how we want that heading to look for molarity.

**Comment by zepumph** [comment]
@terracoda and @emily-phet, I just realized that we currently don't have help content for using the number keys on the "both hands" interaction. Is this something we want to support still? If so, doesn't it deserve keyboard help content?

both hands interaction -> number key 0-9 to move both hands to that value.

**Comment by emily-phet** [comment]
Re the [combobox question](https://github.com/phetsims/ratio-and-proportion/issues/82#issuecomment-712452923) - I have the feeling we'll want different wording related to comboboxes. How standardized do we need to get here? Can there be a default (Molarity's wording seems fine) and then a way to sandwich the combobox name inside of a text string however we want - so we can customize? My guess is this is a situation where the default might get used like 50% of the time, not 90+% so customizing may be more the rule than the exception.

**Comment by emily-phet** [comment]
I made a suggestion re number jumping in the design doc.

**Comment by zepumph** [comment]
@terracoda said I could implement the added row in the design doc about 0-9 in the keyboard dialog. That is all I will do right now.

**Comment by terracoda** [comment]
@zepumph here are some edits:
**Visual Keyboard dialog for screen 1:**
- Change "eight" to right 
- Add new item for 0-9 shortcut
![Screen Shot 2020-10-27 at 4 04 45 PM](https://user-images.githubusercontent.com/1902116/97346693-b3fb6b80-186e-11eb-97c7-3bc94d3380f5.png)

*Accessible keyboard shortcuts screen 1**
- add missing periods
- remove "the" before Shift key
- add new item about number keys
![Screen Shot 2020-10-27 at 4 11 58 PM](https://user-images.githubusercontent.com/1902116/97347262-7e0ab700-186f-11eb-98c7-2a231e64af18.png)

**Comment by terracoda** [comment]
The changes above should also fix the same issues on the second screen's keyboard shortcuts dialog.

I made a decision not to include the Range combobox in dialog on screen 2.

As for comboboxes, in general, it would be good to know how often the default wording will work for learners.

I feel that the numbered list implies ordered steps, so I feel less strongly today about the explicit language we used in Molarity, i.e., "Steps to Change a Solute". Although, in that sim, changing the Solute is primary to the using the simulation, so one could argue for the need for the more explicit instructions.

That said, I think a default pattern could go a long way, and it would be nice to know how often it can actually work for learners. I suggest we make it an option that we can swap out the words in the curly braces:
**{{Choose a {{Challenge Ratio}}}}**
1. Pop up list of {{options}} [Space] or [Enter]
2. Move through {{options}} [Up][Down]
3. Choose new {{option}} [Enter]
4. Close list without changing [Esc]

Here's Molarity's combobox instructions for reference. I would see Molarity as using the customized option because we felt more explicit language was needed.
**{{Steps to Change {{Solute}}}}**
1. Pop up list of {{solutes}} [Space] or [Enter]
2. Move through {{solutes}} [Up][Down]
3. Choose new {{solute}} [Enter]
4. Close list without changing [Esc]

**Steps to Change Solute**
1. Pop up list of solutes with Space or Enter keys.
2. Move through solutes with Up and Down arrow keys.
3. Choose new solute with Enter key.
4. Close list without changing with Escape key.

**Comment by terracoda** [comment]
For the record, the Home screen keyboard dialog looks good to me.

**Comment by terracoda** [comment]
And also, for the record, in https://github.com/phetsims/ratio-and-proportion/issues/82#issuecomment-717475379 I made a decision to not include instructions for the Range combobox because of space issues.

**Comment by terracoda** [comment]
@zepumph, for interviews, we just need to the minor corrections made and the number shortcuts added.

**Comment by zepumph** [comment]
I implemented the bug fixes for RaP. Over to @emily-phet to confirm or discuss the default "Choose a . . . " for combo box help sections from https://github.com/phetsims/ratio-and-proportion/issues/82#issuecomment-717475379.

**Comment by emily-phet** [comment]
From 11/10 meeting - We'd like to remove the left/right indicators from both "move hands" and "move hands in smaller steps".

**Comment by zepumph** [comment]
Sounds good. Can you also comment on https://github.com/phetsims/ratio-and-proportion/issues/82#issuecomment-717560995?

**Comment by zepumph** [comment]
Implemented above. This was all in common code, so we will be able to choose icons for just up/down, just left/right, or both in the future as well. The default is both.

![image](https://user-images.githubusercontent.com/6856943/98741045-727fbb80-2358-11eb-9d8d-643937ec1551.png)


Over to @emily-phet to review and to provide feedback for https://github.com/phetsims/ratio-and-proportion/issues/82#issuecomment-717560995

**Comment by emily-phet** [comment]
I think this looks great! I'm fine with the "choose a challenge ratio" decision.

**Comment by zepumph** [comment]
Sounds great. "Choose a {{thing}}" is the default for ComboBoxHelpSection. Molarity overrides it and says "Steps to Change Solute" Closing

**Comment by terracoda** [comment]
Crispy clean!
